source/_posts/2016-01-25-upgrading-to-ruby-on-rails-5-0.markdown

```diff
@@ -0,0 +1,226 @@
+
+
+## Problem 3 - Model enum attribute not saved
+
+`bin/rails test` gives us ...
+
+```
+Expected to find text "Expense" in "Groceries Edit Delete".
+```
+
+In this test, the Category model, which has a name (in the test the value is Groceries) and a kind with the possible values `income`, `expense` and `any`, comes into play.
+
+The error suggests that the Category is not able to persist the kind value. We quickly try to save a Category with kind expense from the console ...
+
+```
+irb(main):001:0> Category.create!(name: 'Groceries', kind: 'expense')
+=> #<Category id: 1, name: "Groceries", kind: nil, created_at: "2016-01-20 19:29:35", updated_at: "2016-01-20 19:29:35">
+```
+
+There’s where the error comes from, kind isn't saving.
+
+In our case, kind is an enum in the Category model and is represented as a PostgreSQL enum. In order for this combination to work correctly, we need to “hack” the mapping that Active Record creates by default for the enums since, by default, they are returned as integers and we would need Strings for PostgreSQL. Hence, the enums declaration must be something like ... `enum kind: { expense: 'expense', income: 'income', any: 'any' }`. With that declaration, the enums stop using the default mapping to integers and each declared key is mapped to the String that we put in the Hash. This syntax functions correctly in Rails 4.2 but, for some reason, we are not able to save the enums in Rails 5.0.
+
+Debugging, I find out that in Rails 5.0 the deserialization method of the enums (the one that, given a value, returns the enum key), defines `mapping.key(value.to_i)`. This is valid when we don’t customize the mapping. In our case, value is a String and there’s no reason to include `to_i`. Why, then, don't we leave the original value? And [that was precisely the fix](https://github.com/rails/rails/commit/e991c7b8cd69d7ba5e221a19e5f386e3ba02eb9d).
+
+Yet one problem less, but some failing tests remain. 
+
+
+## Problem 4 - Capybara::ElementNotFound: Unable to find button "Create Xxx"
+
+`bin/rails test` gives us ...
+
+```
+Capybara::ElementNotFound: Unable to find button "Create Category"
+```
+
+The error indicates that, for some reason, it can’t find the submit button with the name `Create Category`. A simple look at the page where this button is found shows us that right now the button has the name `Create category`, category in all lowercase.
+
+Looking for changes in the Rails code related to that, I find [this commit](https://github.com/rails/rails/commit/c74f9cc0e8b68697e0d1917b876ffdf404ade3fc). The new form seems fine to me but it’s breaking this type of test. 
+
+We change the tests in our application so that they look through `Create category` then do the same for the rest of the equivalent cases that fail and voila, the tests pass.
+
+With this last change, all of our application’s tests pass. After that, it’s a good idea to test the application by hand and there the following presents itself ...
```

Comment:
`and there the following presents itself` <- doesn't it sound a bit odd?

What about something like the following?
`... it’s a good idea to test the application by hand and then we get the following problem`